common/src/sql/dbinit.sql

```diff
@@ -1002,6 +1004,66 @@ CREATE UNIQUE INDEX ON omicron.public.ip_pool_range (
 STORING (first_address)
 WHERE time_deleted IS NULL;
 
+/*
+ * External IP addresses used for instance source NAT.
+ *
+ * NOTE: This currently stores only address and port information for the
+ * automatic source NAT supplied for all guest instances. It does not currently
+ * store information about ephemeral or floating IPs.
+ */
+CREATE TABLE omicron.public.instance_external_ip (
+    id UUID PRIMARY KEY,
+    time_created TIMESTAMPTZ NOT NULL,
+    time_modified TIMESTAMPTZ NOT NULL,
+    time_deleted TIMESTAMPTZ,
+
+    /* FK to the `ip_pool` table. */
+    ip_pool_id UUID NOT NULL,
+
+    /* FK to the `ip_pool_range` table. */
+    ip_pool_range_id UUID NOT NULL,
+
+    /* FK to the `instance` table. */
+    instance_id UUID NOT NULL,
+
+    /* The actual external IP address. */
+    ip INET NOT NULL,
+
+    /* The first port in the allowed range, inclusive. */
+    first_port INT4 NOT NULL,
+
+    /* The last port in the allowed range, also inclusive. */
+    last_port INT4 NOT NULL
+);
+
+/*
+ * Index used to support quickly looking up children of the IP Pool range table,
+ * when checking for allocated addresses during deletion.
+ */
+CREATE INDEX ON omicron.public.instance_external_ip (
+    ip_pool_id,
+    ip_pool_range_id
+)
+    WHERE time_deleted IS NULL;
+
+/*
+ * Index used to enforce uniqueness of external IPs
+ *
+ * NOTE: This relies on the uniqueness constraint of IP addresses across all
+ * pools, _and_ on the fact that the number of ports assigned to each instance
+ * is fixed at compile time.
+ */
+CREATE UNIQUE INDEX ON omicron.public.instance_external_ip (
+    ip,
+    first_port
+)
+    WHERE time_deleted IS NULL;
+
+CREATE INDEX ON omicron.public.instance_external_ip (
```

Comment:
This is not unique - is that intentional? As in, we expect that instances may have multiple external IPs?

Reply:
That's correct. It's not entirely clear what one would use such an instance for, but something like routing between different external networks might make sense. I think we want to leave it as is for now.